File: common/src/main/kotlin/cn/bakamc/common/config/ConfigUtil.kt
package cn.bakamc.common.config

import cn.bakamc.common.config.modconfig.ConfigCategory
import cn.bakamc.common.utils.getNestedObject
import cn.bakamc.common.utils.getOr
import cn.bakamc.common.utils.gson
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import java.io.*
import java.nio.charset.StandardCharsets
import java.nio.file.Path
import java.util.*

/**
 *

 * 项目名 BakaMC

 * 包名 cn.bakamc.common.config

 * 文件名 ConfigUtil

 * 创建时间 2022/8/12 0:07

 * @author forpleuvoir

 */
object ConfigUtil {

	fun configFile(configFileName: String, path: Path, create: Boolean = true): File {
		val file = File(path.toFile(), "${configFileName}.json")
		if (!file.exists() && create) {
			file.createNewFile()
		}
		return file
	}

	fun writeJsonToFile(root: JsonObject, file: File): Boolean {
		var fileTmp = File(file.parentFile, file.name + ".tmp")
		if (fileTmp.exists()) {
			fileTmp = File(file.parentFile, UUID.randomUUID().toString() + ".tmp")
		}
		try {
			OutputStreamWriter(FileOutputStream(fileTmp), StandardCharsets.UTF_8).use { writer ->
				writer.write(gson.toJson(root))
				writer.close()
				if (file.exists() && file.isFile && !file.delete()) {
					println("Failed to delete file '%s'".format(file.absolutePath))
				}
				return fileTmp.renameTo(file)
			}
		} catch (e: Exception) {
			println("Failed to write JSON data to file '%s'".format(fileTmp.absolutePath))
			e.printStackTrace()
		}
		return false
	}

	fun writeConfigCategory(root: JsonObject, category: ConfigCategory) {
		root.getNestedObject(category.name, true)?.let {
			for (config in category.allConfigs) {
				it.add(config.key, config.serialization)
			}
		}
	}

	fun readConfigCategory(root: JsonObject, category: ConfigCategory) {
		val obj = root.getOr(category.name, JsonObject())
		if (!obj.entrySet().isNullOrEmpty()) {
			category.allConfigs.forEach {
				if (obj.has(it.key)) {
					it.deserialize(obj[it.key])
				}
			}
		}
	}

	fun paresJsonFile(file: File, action: (JsonElement) -> Unit) {
		if (file.exists() && file.isFile && file.canRead()) {
			val fileName = file.absolutePath
			try {
				InputStreamReader(FileInputStream(file), StandardCharsets.UTF_8).use { reader ->
					JsonParser.parseReader(reader).run {
						if (!this.isJsonNull)
							action(this)
					}
				}
			} catch (e: Exception) {
				println("Failed to parse the JSON file '%s'".format(fileName))
				e.printStackTrace()
			}
		}
	}


}

File: nebula-config/src/main/scala/moe/forpleuvoir/nebula/config/ConfigUtil.scala
package moe.forpleuvoir.nebula.config

import moe.forpleuvoir.nebula.common.util.use

import java.io.*
import java.nio.charset.{Charset, StandardCharsets}
import java.nio.file.Path
import java.util.UUID
import scala.io.{Codec, Source}

object ConfigUtil {

  var charset: Charset = StandardCharsets.UTF_8

  def configFile(configFileName: String, path: Path, create: Boolean = true): File = {
    val p = path.toFile
    val f = File(p, configFileName)
    if (!f.exists() && create) {
      if (!p.exists()) {
        p.mkdir()
      }
      try f.createNewFile()
      catch {
        case e: Exception =>
          throw new IOException(s"${e.getMessage},${f.getAbsolutePath}")
      }
    }
    f
  }

  def writeToFile(content: Array[Byte] | String, file: File): Unit = {
    var fileTmp = File(file.getParentFile, file.getName + ".tmp")
    if (fileTmp.exists()) {
      fileTmp = File(file.getParentFile, UUID.randomUUID().toString + ".tmp")
    }

    new FileOutputStream(fileTmp).use { it =>
      content match {
        case c: String =>
          new OutputStreamWriter(it, charset).use { osw =>
            osw.write(c)
          }
        case c: Array[Byte] =>
          it.write(c)
      }
    }
    if (file.exists() && file.isFile && !file.delete()) {
      throw new IOException(s"Failed to delete file ${file.getAbsolutePath}")
    }
    if (!fileTmp.renameTo(file)) {
      throw new IOException(s"Failed to rename temp file to ${file.getAbsolutePath}")
    }
  }

  def readFileToString(file: File): String = {
    if (file.exists() && file.isFile && file.canRead) {
      val is = new FileInputStream(file)
      try Source.fromInputStream(is)(using Codec(charset)).mkString
      finally is.close()
    } else {
      throw new IOException(s"Failed to read the file ${file.getAbsolutePath}")
    }
  }

  def readFile(file: File): Array[Byte] = {
    if (file.exists() && file.isFile && file.canRead) {
      val is = new FileInputStream(file)
      try is.readAllBytes()
      finally is.close()
    } else {
      throw new IOException(s"Failed to read the file ${file.getAbsolutePath}")
    }
  }


}
